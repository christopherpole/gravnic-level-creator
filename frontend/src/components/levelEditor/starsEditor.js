import React from 'react';
import styled, { css } from 'styled-components';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';

import Button from '../common/button';
import StarIcon from '../common/starIcon';

export const Wrapper = styled.div`
  border: 1px solid ${props => props.theme.foregroundColor};
  margin-bottom: ${props => props.theme.structureSpacing};
  padding: ${props => props.theme.structureSpacing};
`;

export const StarsWrapper = styled.ul`
  display: flex;
`;

export const StarIconWrapper = styled.li`
  margin-right: 0.2em;
  list-style-type: none;

  &:last-child {
    margin-right: 0;
  }
`;

export const StarsList = styled.ul``;

export const StarListItem = styled.li`
  display: flex;
  justify-content: space-between;
  list-style-type: none;
  align-items: center;
  margin-bottom: 0.3em;

  &:last-child {
    margin-bottom: 0;
  }
`;

export const StarsLabel = styled.span`
  font-size: 1.2em;
`;

export const ControlsWrapper = styled.div`
  display: flex;
`;

export const ControlButton = styled(Button)`
  margin-right: 0.3em;
  padding: 0;
  height: 2em;
  width: 2em;
  line-height: 2em;

  &:last-child {
    margin-right: 0;
  }
`;

export const StarsEditor = () => (
  <Wrapper id="stars-editor">
    <StarsList>
      {[...Array(3)].map((_, i) => (
        <StarListItem key={i}>
          <StarsWrapper>
            {[...Array(3)].map((meh, j) => (
              <StarIconWrapper key={j}>
                <StarIcon fillColor={i > j ? 'transparent' : 'white'} size={20} />
              </StarIconWrapper>
            ))}
          </StarsWrapper>

          <StarsLabel>99</StarsLabel>

          <ControlsWrapper>
            <ControlButton>-</ControlButton>
            <ControlButton>+</ControlButton>
          </ControlsWrapper>
        </StarListItem>
      ))}
    </StarsList>
  </Wrapper>
);

StarsEditor.defaultProps = {};

StarsEditor.propTypes = {};

const mapStateToProps = state => ({});

const mapDispatchToProps = dispatch => ({});

export default connect(mapStateToProps, mapDispatchToProps)(StarsEditor);
