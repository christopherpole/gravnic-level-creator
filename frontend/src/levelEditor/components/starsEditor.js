import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';

import Button from 'common/button';
import StarIcon from 'common/starIcon';
import { setStars } from '../actions';

export const Wrapper = styled.div`
  border: 1px solid ${props => props.theme.foregroundColor};
  margin-bottom: ${props => props.theme.structureSpacing};
  padding: ${props => props.theme.structureSpacing};
`;

export const StarsWrapper = styled.ul`
  display: flex;
`;

export const StarIconWrapper = styled.li`
  margin-right: 0.2em;
  list-style-type: none;

  &:last-child {
    margin-right: 0;
  }
`;

export const StarsList = styled.ul``;

export const StarListItem = styled.li`
  display: flex;
  justify-content: space-between;
  list-style-type: none;
  align-items: center;
  margin-bottom: 0.3em;

  &:last-child {
    margin-bottom: 0;
  }
`;

export const StarsLabel = styled.span`
  font-size: 1.2em;
  user-select: none;
`;

export const ControlsWrapper = styled.div`
  display: flex;
`;

export const ControlButton = styled(Button)`
  margin-right: 0.3em;
  padding: 0;
  height: 2em;
  width: 2em;
  line-height: 2em;

  &:last-child {
    margin-right: 0;
  }
`;

export const ButtonDecrement = styled(ControlButton)``;

export const ButtonIncrement = styled(ControlButton)``;

export const StarsEditor = ({ stars, setStarsAction }) => (
  <Wrapper id="stars-editor">
    <StarsList>
      {stars.map((noOfMoves, i) => (
        <StarListItem key={i}>
          <StarsWrapper>
            {[...Array(3)].map((_, j) => (
              <StarIconWrapper key={j}>
                <StarIcon fillColor={i > j ? 'transparent' : undefined} size={20} />
              </StarIconWrapper>
            ))}
          </StarsWrapper>

          <StarsLabel>{noOfMoves}</StarsLabel>

          <ControlsWrapper>
            <ButtonDecrement
              onClick={() => {
                setStarsAction(i, noOfMoves - 1);
              }}
              className="btn-decrement"
            >
              -
            </ButtonDecrement>
            <ButtonIncrement
              onClick={() => {
                setStarsAction(i, noOfMoves + 1);
              }}
              className="btn-increment"
            >
              +
            </ButtonIncrement>
          </ControlsWrapper>
        </StarListItem>
      ))}
    </StarsList>
  </Wrapper>
);

StarsEditor.defaultProps = {};

StarsEditor.propTypes = {
  stars: PropTypes.array.isRequired,
  setStarsAction: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  stars: state.levelEditor.stars,
});

const mapDispatchToProps = dispatch => ({
  setStarsAction: bindActionCreators(setStars, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(StarsEditor);
