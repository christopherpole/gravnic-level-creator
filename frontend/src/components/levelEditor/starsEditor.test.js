import React from 'react';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { spy } from 'sinon';

import { StarsEditor } from './starsEditor';

configure({ adapter: new Adapter() });

const props = {};

describe('The tile selector', () => {
  it('Renders without exploding', () => {
    const starsEditor = shallow(<StarsEditor {...props} />);

    expect(starsEditor).toHaveLength(1);
  });

  it('Matches the current snapshot', () => {
    const starsEditor = shallow(<StarsEditor {...props} />);

    expect(toJson(starsEditor)).toMatchSnapshot();
  });
});
