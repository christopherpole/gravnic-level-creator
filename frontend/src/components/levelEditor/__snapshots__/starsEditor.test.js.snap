// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The tile selector Matches the current snapshot 1`] = `
<styled.div
  id="stars-editor"
>
  <styled.ul>
    <styled.li
      key="0"
    >
      <styled.ul>
        <styled.li
          key="0"
        >
          <StarIcon
            fillColor="white"
            size={20}
          />
        </styled.li>
        <styled.li
          key="1"
        >
          <StarIcon
            fillColor="white"
            size={20}
          />
        </styled.li>
        <styled.li
          key="2"
        >
          <StarIcon
            fillColor="white"
            size={20}
          />
        </styled.li>
      </styled.ul>
      <styled.span>
        99
      </styled.span>
      <styled.div>
        <Styled(Button)>
          -
        </Styled(Button)>
        <Styled(Button)>
          +
        </Styled(Button)>
      </styled.div>
    </styled.li>
    <styled.li
      key="1"
    >
      <styled.ul>
        <styled.li
          key="0"
        >
          <StarIcon
            fillColor="transparent"
            size={20}
          />
        </styled.li>
        <styled.li
          key="1"
        >
          <StarIcon
            fillColor="white"
            size={20}
          />
        </styled.li>
        <styled.li
          key="2"
        >
          <StarIcon
            fillColor="white"
            size={20}
          />
        </styled.li>
      </styled.ul>
      <styled.span>
        99
      </styled.span>
      <styled.div>
        <Styled(Button)>
          -
        </Styled(Button)>
        <Styled(Button)>
          +
        </Styled(Button)>
      </styled.div>
    </styled.li>
    <styled.li
      key="2"
    >
      <styled.ul>
        <styled.li
          key="0"
        >
          <StarIcon
            fillColor="transparent"
            size={20}
          />
        </styled.li>
        <styled.li
          key="1"
        >
          <StarIcon
            fillColor="transparent"
            size={20}
          />
        </styled.li>
        <styled.li
          key="2"
        >
          <StarIcon
            fillColor="white"
            size={20}
          />
        </styled.li>
      </styled.ul>
      <styled.span>
        99
      </styled.span>
      <styled.div>
        <Styled(Button)>
          -
        </Styled(Button)>
        <Styled(Button)>
          +
        </Styled(Button)>
      </styled.div>
    </styled.li>
  </styled.ul>
</styled.div>
`;
